<a href="https://colab.research.google.com/github/fatemenoroozi/color_images/blob/main/color_images.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


Import libraries

In [1]:
import cv2
import matplotlib.pyplot as plt
import random
import numpy as np
import os
from PIL import Image
import os

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers,models
from tensorflow.keras.models import Sequential

##Create color images

In [2]:
def create_color_images(width, height,color):
  image = np.zeros((height, width, 3), np.uint8)
  image[:] = color
  return image

Create directory

In [3]:
dataset_folder_name = 'dataset'
red_image_folder_name = 'red'
green_image_folder_name = 'green'
blue_image_folder_name = 'blue'
yellow_image_folder_name = 'yellow'
black_image_folder_name = 'black'
white_image_folder_name = 'white'
purple_image_folder_name = 'purple'
red_image_folder_path = os.path.join(dataset_folder_name, red_image_folder_name)
green_image_folder_path = os.path.join(dataset_folder_name, green_image_folder_name)
blue_image_folder_path = os.path.join(dataset_folder_name, blue_image_folder_name)
yellow_image_folder_path = os.path.join(dataset_folder_name, yellow_image_folder_name)
black_image_folder_path = os.path.join(dataset_folder_name, black_image_folder_name)
white_image_folder_path = os.path.join(dataset_folder_name, white_image_folder_name)
purple_image_folder_path = os.path.join(dataset_folder_name, purple_image_folder_name)
os.makedirs(red_image_folder_path)
os.makedirs(green_image_folder_path)
os.makedirs(blue_image_folder_path)
os.makedirs(yellow_image_folder_path)
os.makedirs(black_image_folder_path)
os.makedirs(white_image_folder_path)
os.makedirs(purple_image_folder_path)

In [4]:
width = 180
height = 180

Red images

In [5]:
for i in range (1000):
  color = (0,0,random.randint(50,255))
  image = create_color_images(width, height,color)
  file_name = f"red_{i}.png"
  path = "dataset/red"
  cv2.imwrite(os.path.join(path , file_name), image)

Green images

In [6]:
for i in range (1000):
  color = (0,random.randint(50,255),0)
  image = create_color_images(width, height,color)
  file_name = f"green_{i}.png"
  path = "dataset/green"
  cv2.imwrite(os.path.join(path , file_name), image)

Blue images

In [7]:
for i in range (1000):
  color = (random.randint(50,255),0,0)
  image = create_color_images(width, height,color)
  file_name = f"blue_{i}.png"
  path = "dataset/blue"
  cv2.imwrite(os.path.join(path , file_name), image)

Purple images

In [8]:
for i in range (1000):
  color = (random.randint(100,135),0,random.randint(100,135))
  image = create_color_images(width, height,color)
  file_name = f"purple_{i}.png"
  path = "dataset/purple"
  cv2.imwrite(os.path.join(path , file_name), image)

Yellow images

In [9]:
for i in range (1000):
  color = (0,random.randint(230,255),random.randint(230,255))
  image = create_color_images(width, height,color)
  file_name = f"yellow_{i}.png"
  path = "dataset/yellow"
  cv2.imwrite(os.path.join(path , file_name), image)

Black images

In [10]:
for i in range (500):
  color = (random.randint(0,15),random.randint(0,15),random.randint(0,15))
  image = create_color_images(width, height,color)
  file_name = f"black_{i}.png"
  path = "dataset/black"
  cv2.imwrite(os.path.join(path , file_name), image)

white image

In [11]:
for i in range (500):
  color = (random.randint(250,255),random.randint(250,255),random.randint(250,255))
  image = create_color_images(width, height,color)
  file_name = f"white_{i}.png"
  path = "dataset/white"
  cv2.imwrite(os.path.join(path , file_name), image)

In [13]:
directory = 'dataset'

train_ds , val_ds = tf.keras.utils.image_dataset_from_directory(
    directory,
    color_mode='rgb',
    batch_size=32,
    image_size=(180, 180),
    validation_split=0.3,
    seed = 123,
    subset="both"
)

Found 6000 files belonging to 7 classes.
Using 4200 files for training.
Using 1800 files for validation.


In [15]:
class_names = train_ds.class_names
class_names

['black', 'blue', 'green', 'purple', 'red', 'white', 'yellow']

Create the model

In [16]:
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(180, 180, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(7))

In [17]:
model.compile(
  optimizer='adam',
  loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
  metrics=['accuracy'])

In [18]:
epochs= 5
history = model.fit(
  train_ds,
  validation_data=val_ds,
  epochs=epochs
)

Epoch 1/5
132/132 [==============================] - 19s 47ms/step - loss: 3.1931 - accuracy: 0.8552 - val_loss: 37.7033 - val_accuracy: 0.7350
Epoch 2/5
132/132 [==============================] - 5s 37ms/step - loss: 1.6035 - accuracy: 0.9183 - val_loss: 1.1421 - val_accuracy: 0.9494
Epoch 3/5
132/132 [==============================] - 6s 48ms/step - loss: 0.8305 - accuracy: 0.9626 - val_loss: 0.9968 - val_accuracy: 0.9250
Epoch 4/5
132/132 [==============================] - 5s 39ms/step - loss: 0.4205 - accuracy: 0.9731 - val_loss: 0.1754 - val_accuracy: 0.9928
Epoch 5/5
132/132 [==============================] - 6s 43ms/step - loss: 0.0701 - accuracy: 0.9962 - val_loss: 0.0118 - val_accuracy: 0.9978


the model has achieved validation accuracy of over 99%

##Noisy images

Create directory

In [20]:
dataset_folder_name = 'dataset'
noisy_dataset_folder_name = 'noisy_dataset'
noisy_green_image_folder_name = 'noisy_green'
noisy_red_image_folder_name = 'noisy_red'
noisy_blue_image_folder_name = 'noisy_blue'
noisy_purple_image_folder_name = 'noisy_purple'
noisy_white_image_folder_name = 'noisy_white'
noisy_black_image_folder_name = 'noisy_black'
noisy_yellow_image_folder_name = 'noisy_yellow'
noisy_green_folder_path = os.path.join(noisy_dataset_folder_name , noisy_green_image_folder_name)
noisy_red_folder_path = os.path.join(noisy_dataset_folder_name , noisy_red_image_folder_name)
noisy_blue_folder_path = os.path.join(noisy_dataset_folder_name , noisy_blue_image_folder_name)
noisy_purple_folder_path = os.path.join(noisy_dataset_folder_name , noisy_purple_image_folder_name)
noisy_white_folder_path = os.path.join(noisy_dataset_folder_name , noisy_white_image_folder_name)
noisy_black_folder_path = os.path.join(noisy_dataset_folder_name , noisy_black_image_folder_name)
noisy_yellow_folder_path = os.path.join(noisy_dataset_folder_name , noisy_yellow_image_folder_name)
os.makedirs(noisy_green_folder_path)
os.makedirs(noisy_red_folder_path)
os.makedirs(noisy_blue_folder_path)
os.makedirs(noisy_purple_folder_path)
os.makedirs(noisy_white_folder_path)
os.makedirs(noisy_black_folder_path)
os.makedirs(noisy_yellow_folder_path)

Noisy green images

In [21]:
folder_path = "/content/dataset/green"
for filename in os.listdir(folder_path):
  if filename.endswith('.png'):
    img = cv2.imread(os.path.join(folder_path,filename))
    noise = np.random.normal(0,70,(180,180,3))
    noisy_image = img + noise
    path = '/content/noisy_dataset/noisy_green/'
    cv2.imwrite(os.path.join(path , 'noisy_' + filename), noisy_image)

Noisy red images

In [22]:
folder_path = "/content/dataset/red"
for filename in os.listdir(folder_path):
  if filename.endswith('.png'):
    img = cv2.imread(os.path.join(folder_path,filename))
    noise = np.random.normal(0,70,(180,180,3))
    noisy_image = img + noise
    path = '/content/noisy_dataset/noisy_red/'
    cv2.imwrite(os.path.join(path , 'noisy_' + filename), noisy_image)

Noisy blue images

In [23]:
folder_path = "/content/dataset/blue"
for filename in os.listdir(folder_path):
  if filename.endswith('.png'):
    img = cv2.imread(os.path.join(folder_path,filename))
    noise = np.random.normal(0,70,(180,180,3))
    noisy_image = img + noise
    path = '/content/noisy_dataset/noisy_blue/'
    cv2.imwrite(os.path.join(path , 'noisy_' + filename), noisy_image)

Noisy purple images

In [24]:
folder_path = "/content/dataset/purple"
for filename in os.listdir(folder_path):
  if filename.endswith('.png'):
    img = cv2.imread(os.path.join(folder_path,filename))
    noise = np.random.normal(0,70,(180,180,3))
    noisy_image = img + noise
    path = '/content/noisy_dataset/noisy_purple/'
    cv2.imwrite(os.path.join(path , 'noisy_' + filename), noisy_image)

Noisy yellow images

In [25]:
folder_path = "/content/dataset/yellow"
for filename in os.listdir(folder_path):
  if filename.endswith('.png'):
    img = cv2.imread(os.path.join(folder_path,filename))
    noise = np.random.normal(0,70,(180,180,3))
    noisy_image = img + noise
    path = '/content/noisy_dataset/noisy_yellow/'
    cv2.imwrite(os.path.join(path , 'noisy_' + filename), noisy_image)

Noisy white images

In [26]:
folder_path = "/content/dataset/white"
for filename in os.listdir(folder_path):
  if filename.endswith('.png'):
    img = cv2.imread(os.path.join(folder_path,filename))
    noise = np.random.normal(0,70,(180,180,3))
    noisy_image = img + noise
    path = '/content/noisy_dataset/noisy_white/'
    cv2.imwrite(os.path.join(path , 'noisy_' + filename), noisy_image)

Noisy black images

In [27]:
folder_path = "/content/dataset/black"
for filename in os.listdir(folder_path):
  if filename.endswith('.png'):
    img = cv2.imread(os.path.join(folder_path,filename))
    noise = np.random.normal(0,70,(180,180,3))
    noisy_image = img + noise
    path = '/content/noisy_dataset/noisy_black/'
    cv2.imwrite(os.path.join(path , 'noisy_' + filename), noisy_image)

Create a dataset

In [28]:
directory = 'noisy_dataset'

train_ds , val_ds = tf.keras.utils.image_dataset_from_directory(
    directory,
    color_mode='rgb',
    batch_size=32,
    image_size=(180, 180),
    validation_split=0.2,
    seed = 123,
    subset="both"
)

Found 6000 files belonging to 7 classes.
Using 4800 files for training.
Using 1200 files for validation.


Create the model

In [29]:
model = models.Sequential()
model.add(layers.Conv2D(32 , (3, 3), padding='same', activation='relu', input_shape=(180, 180, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(32, (3, 3), padding= 'same',activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), padding='same' , activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(7))

In [30]:
model.compile(
  optimizer='adam',
  loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
  metrics=['accuracy'])

In [31]:
epochs=5
history = model.fit(
  train_ds,
  validation_data=val_ds,
  epochs=epochs
)

Epoch 1/5
150/150 [==============================] - 13s 68ms/step - loss: 5.9244 - accuracy: 0.7633 - val_loss: 0.6285 - val_accuracy: 0.8408
Epoch 2/5
150/150 [==============================] - 10s 67ms/step - loss: 0.6087 - accuracy: 0.8315 - val_loss: 0.5945 - val_accuracy: 0.8408
Epoch 3/5
150/150 [==============================] - 9s 60ms/step - loss: 0.5759 - accuracy: 0.8315 - val_loss: 0.5763 - val_accuracy: 0.8383
Epoch 4/5
150/150 [==============================] - 11s 67ms/step - loss: 0.5548 - accuracy: 0.8304 - val_loss: 0.5343 - val_accuracy: 0.8408
Epoch 5/5
150/150 [==============================] - 10s 68ms/step - loss: 0.3849 - accuracy: 0.8171 - val_loss: 0.2551 - val_accuracy: 0.8408


the model whit noisy data has achieved validation accuracy of over 84%